### ISSUE TITLE
Running native tests xplat

### ISSUE THREAD

**Issue by jaredwy** [Microsoft/ChakraCore#2590]
In an attempt to write some unit tests for #2581 I hit a slight snag. The native tests that test the JSRT api is driven by a windows only build. 

I managed to get the driver to compile, this took a few #defines and an update to catch.hpp. Also I had to use libc++ as the version of stdlib++ doesn't seem to support some of the features that are needed by catch. I would need to look on linux to see what happens here.

The other problem is the rather liberal use of windows only jsrt functions. E.g.

```
REQUIRE(JsGetPropertyNameFromId(name1, &name) == JsNoError); REQUIRE(JsPointerToString(_u("value1"), wcslen(_u("value1")), &value1) == JsNoError);
```

My suggestion is rewrite these to use `JsCopyString` etc. If their is still a need to test JsPointerToString this can be done with platform specific test files.

**Comment by digitalinfinity** [comment]
Thanks @jaredwy - talked offline with @curtisman and we don't see any problems moving these tests to the new Chakra API path- we would welcome a PR here 😄 

sidenote- curious, what was the issue with catch and stdlib++ (cc @akroshg)

**Comment by jaredwy** [comment]
Happy to spin up a PR. Ill have some time this weekend.

Any issues with MS just dropping in the latest version of catch from github? 
I may need a hand just checking the windows side of things, as I don't have a windows machine at present. 


I don't think it is a problem with stdlib++ per say, just an issue with the version that ships in osx. I was getting a bunch of missing symbol errors. I didn't look too far into it. Been on mac long enough to know that there are issues with gnu tooling :)

**Comment by akroshg** [comment]
@jaredwy I haven't dropped the latest version of the catch. You are welcome to try with the new version.

**Comment by digitalinfinity** [comment]
@jaredwy happy to help with the windows side of things- you can share your branch here (or if you open a PR against our master branch, our CI will run the Windows and Linux builds automatically anyway)

**Comment by jaredwy** [comment]
Are these run on linux?  I couldn't find any evidence that they were.

**Comment by digitalinfinity** [comment]
No, the native tests are not run on Linux (they will be after your PR hopefully 😄). The entry point to invoke them on Windows is test\runnativetests.cmd and I believe they run in our Jenkins CI (@dilijev/@akroshg, correct me if I'm wrong here)

From your PR perspective, I think you'll need to do the following:
- Switch our native unit tests to run using the xplat Jsrt APIs (or better still, two-path it so that on Linux/macOs, it tests the xplat APIs, and on Windows, it tests both the old APIs and the new xplat ones)
- Add a driver similar to runnativetests.cmd - you could just add this to either runtests.sh or runtests.py

I think if you do those two steps, then the CI should help catch any issues from your change. Thanks again for offering to make this change- we really appreciate the contribution!

**Comment by jaredwy** [comment]
No problem.
I will see what makes sense. I will also probably need to add the building of the executable to cmake. Might make it standalone than attempting to wire it into the rest of the cmake infrastructure for now.

**Comment by obastemur** [comment]
@digitalinfinity we already have native tests setup for xplat. https://github.com/Microsoft/ChakraCore/tree/master/test/native-tests 

@jaredwy You may just create another folder there and copy the `Platform.js` from one of the folders. After that;

- introduce the source files under https://github.com/Microsoft/ChakraCore/blob/master/test/native-tests/test-static-native/Platform.js#L44 

- add new test folder similar to this -> https://github.com/Microsoft/ChakraCore/blob/master/test/native-tests/test_native.sh#L88

To test your file on your local

You need to compile ChakraCore once in one of the Test/Debug/Release mode (static build). After that, go under test/native-tests and;
```
$> ./test_native.sh <path to ch> <build_type -> Test/Debug/Release>
```

This will create a Makefile for each test folders, you may visit your folder and continue your experiment there by calling `make` under that folder for future experiments.

**Comment by digitalinfinity** [comment]
@obastemur why do we have a different set of native tests for xplat vs Windows? Should we combine the two? Seems like if @jaredwy moves NativeUnitTests to xplat, it should be fairly straightforward to merge that with native-tests?

**Comment by obastemur** [comment]
Agree we better have one native test approach across all the platforms. My choice would be the xplat native tests approach.

Current xplat native tests are individual applications. We need them to be that way to test various build alternatives (in addition to feature tests) 

So, we test both embedder side and API

We may use xplat tests approach for Windows as well and combine under tests/native-tests

**Comment by jaredwy** [comment]
@obastemur those tests are rather limited in what they can achieve as it stands. They simply compile and execute a basic file and check for a success/failure message and the return code.

While this assures that the apis are exposed there is nothing in them to the various apis behave correctly. test_native.sh would have to be extended to read the output from catch.

**Comment by obastemur** [comment]
@jaredwy They are unit tests, designed to do one thing at a time but that doesn't mean the subsystem can't be used for more complicated tests.

- Makefile is automated from Platform.js

- Test is executed by test_native.sh

You will endup creating a makefile and and sh connection, why do the work from scratch ? I really didn't understand what is better if you do this under bin/... folder?

You do not have to expose and compare the messages from test_native.sh. Whatever you were doing under bin/ folder, you will do the same under test/native-tests;

but It will be a less work.

In order to comply with test_native.sh, you may simply print a basic `PASSED` etc. text at the end of the tests. or add a new message similar to https://github.com/Microsoft/ChakraCore/blob/master/test/native-tests/test_native.sh#L59

**Comment by obastemur** [comment]
BTW; I'm around during the weekend, ping me from `gitter` in case you want to chat.

**Comment by jaredwy** [comment]
I at the mercy of catch.hpp output as it stands.
Is your intention to remove the dependency on catch?

**Comment by jaredwy** [comment]
It's easy enough in the passed case but it certainly helps dev speed to have easy access to output of the various asserts from catch. So would probably have to dump stdout/err on failure.

**Comment by obastemur** [comment]
> So would probably have to dump stdout/err on failure. 

Dev can visit the folder and call `make;./sample.o` and get the output.

**Comment by obastemur** [comment]
Also see https://github.com/Microsoft/ChakraCore/blob/master/test/native-tests/test_native.sh#L30 for dumping errors.

**Comment by jaredwy** [comment]
Just an update.

I have the majority of the tests inside `JsRTApiTest.cpp` compiling and passing on mac. Im working on the tests that require reading from a file at the moment and just chasing down getting palrt compiling :) (For some reason, clang reports it self as gnuc
```
$ clang -dM -E -x c /dev/null | grep -i gnu 
#define __GNUC_MINOR__ 2
#define __GNUC_PATCHLEVEL__ 1
#define __GNUC_STDC_INLINE__ 1
#define __GNUC__ 4
```
 but doesn't provide __declytpe.

Which means that 
```
#if defined(_MSC_VER) || (defined(__GNUC__) && (__GNUC__ >= 3))
#define SAFECRT_DEPRECATED __declspec(deprecated)
#else
#define SAFECRT_DEPRECATED
#endif
```

takes the wrong path. This compiles fine in ch so I just need to fiddle with compiler settings until it works :)

**Comment by jaredwy** [comment]
Ok i have got the majority of tests working on mac (it should in theory work for linux as well) over at 
https://github.com/jaredwy/ChakraCore/tree/update-native-tests-xplat 

It needs some cleaning up, but wanted to get them working to start a discussion on the approach before i start moving tests around.

Some of the tests are commented out until I can figure out the cross plat apis. The documentation for the seralize calls seems out of line with the headers, same with modules.

The biggest concern right now is string handling @obastemur mentioned they prefer utf for the strings in tests. While this encodes the strings as utf-8 some of the tests checking lengths will fail if we use anything outside of ascii. Is this acceptable? 

I can add tests for the windows only functions `JsCopyString` etc in a windows only file. 

I added some convenience methods to deal with everything as std::strings, and also mirrored some of the windows apis like JsRunScript is this acceptable or should i just stick to using char *?  The plan here will be to run the tests twice on windows. Once with the `JsRunScript` that calls `JsRun` and then again with just JsRunScript. Will add a bit of time to your test run but you get extra coverage with out needing to write two sets of tests :)

I moved some #defines from pal into here. Due to the issue above, I have yet to get pal compiling, so I just created a very simple file loader rather than attempting to re-use the existing one. 
It will have to be a little more complex in the future to be able to test JsParseScriptAttributeArrayBufferIsUtf16Encoded with loaded scripts. 

I will try and be on gitter tomorrow to flesh out the answers to these so I can the branch ready for a PR.

**Comment by obastemur** [comment]
> While this encodes the strings as utf-8 some of the tests checking lengths will fail if we use anything outside of ascii. Is this acceptable?

What do you mean by this? an example?

`#define SAFECRT_DEPRECATED __declspec(deprecated)`

Under pal_mstypes.h -> 

```
#if !defined(_declspec)
#define _declspec(e)  __declspec(e)
#endif
```

So, it always takes that path and doesn't fail.

BTW; I saw you also updated Catch. Can we do it without updating?

**Comment by obastemur** [comment]
```
#ifdef __GNUC__
#define SAFECRT_DEPRECATED
#endif
```

this is also fine.

**Comment by jaredwy** [comment]
> BTW; I saw you also updated Catch. Can we do it without updating?

Nope, the version in the code base wouldn't compile using clang on osx seirra. 

> What do you mean by this? an example?
the easiest way to show it is by example http://cpp.sh/7pl65. Hit run and look at the output. Length etc count bytes not characters. We need a utf-8 aware container if we want to support non ascii chars :) and allow people to count the characters in a test. 


pal_mstypes.h is a preprocessor macro so it will work, this fails attempting to compile the expansions of SAFECRT_DEPRECATED. I am not comfortable making changes inside that code as it does seem to compile on osx, i just must be missing something. It may not be important, if we are happy not to use the old filereader.

**Comment by digitalinfinity** [comment]
@jaredwy any update on this issue? Are you still interested in working on this?

**Comment by am11** [comment]
Can we try setting up platform-agnostic [ctest](https://cmake.org/Wiki/CMake/Testing_With_CTest)? We will achieve something like:

```cmd
:: clone
git clone https://Microsoft/ChakraCore cc --recusive
cd cc

:: build
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
cmake --build . --config Release

:: test
ctest
```

**Comment by jaredwy** [comment]
More than happy to keep working on this. I had a branch that got the tests passing on mac. I had some questions on how to best split the tests for the platform specific (windows only) apis. The handling of strings etc.  I think people were heads down getting 2.0 so struggled to find answers. I can find some time in the coming days to refresh my memory and get some answers to those questions. 

Who would be best to chat to about this?

**Comment by obastemur** [comment]
@jaredwy let me invite you to our cross platform channel. Most of the related ppl are on that channel. 
@digitalinfinity @jianchun can you send the invite pls?

BTW; you can also ping from private. IIRC; we had a discussion going on. Thanks for your time! 👍
